Validation/RecoParticleFlow/plugins/PFCaloGPUComparisonTask.cc

```diff
@@ -0,0 +1,170 @@
+      pfCaloGPUCompDir{conf.getUntrackedParameter<std::string>("pfCaloGPUCompDir")} {}
+
+void PFCaloGPUComparisonTask::bookHistograms(DQMStore::IBooker& ibooker,
+                                             edm::Run const& irun,
+                                             edm::EventSetup const& isetup) {
+  const char* histo;
+
+  ibooker.setCurrentFolder("ParticleFlow/" + pfCaloGPUCompDir);
+
+  histo = "pfCluster_Multiplicity_GPUvsCPU_";
+  pfCluster_Multiplicity_GPUvsCPU_ = ibooker.book2D(histo, histo, 100, 0, 2000, 100, 0, 2000);
+
+  histo = "pfCluster_Energy_GPUvsCPU_";
+  pfCluster_Energy_GPUvsCPU_ = ibooker.book2D(histo, histo, 100, 0, 500, 100, 0, 500);
+
+  histo = "pfCluster_RecHitMultiplicity_GPUvsCPU_";
+  pfCluster_RecHitMultiplicity_GPUvsCPU_ = ibooker.book2D(histo, histo, 100, 0, 100, 100, 0, 100);
+
+  histo = "pfCluster_Layer_GPUvsCPU_";
+  pfCluster_Layer_GPUvsCPU_ = ibooker.book2D(histo, histo, 100, 0, 100, 100, 0, 100);
+
+  histo = "pfCluster_Depth_GPUvsCPU_";
+  pfCluster_Depth_GPUvsCPU_ = ibooker.book2D(histo, histo, 100, 0, 100, 100, 0, 100);
+
+  histo = "pfCluster_Eta_GPUvsCPU_";
+  pfCluster_Eta_GPUvsCPU_ = ibooker.book2D(histo, histo, 100, 0, 100, 100, 0, 100);
+
+  histo = "pfCluster_Phi_GPUvsCPU_";
+  pfCluster_Phi_GPUvsCPU_ = ibooker.book2D(histo, histo, 100, 0, 100, 100, 0, 100);
+}
+void PFCaloGPUComparisonTask::analyze(edm::Event const& event, edm::EventSetup const& c) {
+  edm::Handle<reco::PFClusterCollection> pfClusters_ref;
+  event.getByToken(pfClusterTok_ref_, pfClusters_ref);
+
+  edm::Handle<reco::PFClusterCollection> pfClusters_target;
+  event.getByToken(pfClusterTok_target_, pfClusters_target);
+
+  //
+  // Compare per-event PF cluster multiplicity
+
+  if (pfClusters_ref->size() != pfClusters_target->size())
+    LOGVERB("PFCaloGPUComparisonTask") << " PFCluster multiplicity " << pfClusters_ref->size() << " "
+                                       << pfClusters_target->size();
+  pfCluster_Multiplicity_GPUvsCPU_->Fill((float)pfClusters_ref->size(), (float)pfClusters_target->size());
+
+  //
+  // Find matching PF cluster pairs
+  std::vector<int> matched_idx;
+  for (unsigned i = 0; i < pfClusters_ref->size(); ++i) {
+    bool matched = false;
+    for (unsigned j = 0; j < pfClusters_target->size(); ++j) {
+      if (pfClusters_ref->at(i).seed() == pfClusters_target->at(j).seed()) {
+        if (!matched) {
+          matched = true;
+          matched_idx.push_back((int)j);
+        } else {
+          edm::LogWarning("PFCaloGPUComparisonTask") << " another matching? ";
+        }
+      }
+    }
+    if (!matched)
+      matched_idx.push_back(-1);  // if you don't find a match, put a dummy number
+  }
+
+  //
+  // Check matches
+  std::vector<int> tmp = matched_idx;
+  sort(tmp.begin(), tmp.end());
+  const bool hasDuplicates = std::adjacent_find(tmp.begin(), tmp.end()) != tmp.end();
+  if (hasDuplicates)
+    edm::LogWarning("PFCaloGPUComparisonTask") << "find duplicated matches";
```

Comment:
Would a change like this make sense (assuming earlier the histogram is booked)?
@hatakeyamak @fwyzard 
```suggestion
  // Find matching PF cluster pairs
  std::vector<int> matched_idx;
  std::vector<int> duplicate_idx;
  for (unsigned i = 0; i < pfClusters_ref->size(); ++i) {
    bool matched = false;
    for (unsigned j = 0; j < pfClusters_target->size(); ++j) {
      if (pfClusters_ref->at(i).seed() == pfClusters_target->at(j).seed()) {
        if (!matched) {
          matched = true;
          matched_idx.push_back((int)j);
        } else {
          edm::LogWarning("PFCaloGPUComparisonTask") << "Found duplicate match";
          duplicate_idx.push_back((int)j);
        }
      }
    }
    if (!matched)
      matched_idx.push_back(-1);  // if you don't find a match, put a dummy number
  }

  //
  // Plot duplicate matches
  if (!duplicate_idx.empty()) {
    for (unsigned i : duplicate_idx) {
      pfCluster_DuplicateMatches_GPUvsCPU_->Fill(i);
    }
  }
```

Reply:
Right, I should just fill directly.